packages/@aws-cdk/cx-api/lib/assembly/cloud-assembly.ts

```diff
@@ -0,0 +1,105 @@
+import fs = require('fs');
+import path = require('path');
+import semver = require('semver');
+import { AppRuntime, AssemblyManifest, MANIFEST_FILE, MissingContext, PROTO_RESPONSE_VERSION } from '../cxapi';
+import { CloudArtifact } from './cloud-artifact';
+import { ICloudArtifact, ICloudAssembly, ICloudFormationStackArtifact } from './cloud-assembly-api';
+import { CloudFormationStackArtifact } from './cloudformation-stack';
+import { topologicalSort } from './toposort';
+
+export class CloudAssembly implements ICloudAssembly {
+  public readonly artifacts: ICloudArtifact[];
+  public readonly version: string;
+  public readonly missing?: { [key: string]: MissingContext };
+  public readonly runtime: AppRuntime;
+
+  private readonly manifest: AssemblyManifest;
+
+  constructor(public readonly directory: string) {
+    this.manifest = this.readJson(MANIFEST_FILE);
+
+    this.version = this.manifest.version;
+    verifyManifestVersion(this.manifest.version);
+
+    this.artifacts = this.renderArtifacts();
+    this.missing = this.renderMissing();
+    this.runtime = this.manifest.runtime || { libraries: { } };
+
+    // force validation of deps by accessing 'depends' on all artifacts
+    this.validateDeps();
+  }
+
+  public readJson(filePath: string) {
```

Comment:
That looks like a very low level API - who is it public? It looks half-baked or very inconvenient to use. The result is `any` - that's not great.